### PR TITLE
Fbl scheme

### PR DESCRIPTION
**Pull request by setmar**
Implemented backward-forward scheme with wall boundaries, based on python reference code.

To test simulation:
Build and run
./main --etaNo=0 --bathymetryNo=0 --nx=256 --ny 256 --width=100 --height=100 --wallDuration=10 --outputFile=test.nc

### CODE REVIEW

File: doc/reference_python/SWESimulators/CTCS.py [hunk 1]
```diff
@@ -60,14 +61,24 @@ class CTCS:
...
-                 cl_ctx, \
+                 #cl_ctx, \
                  H, eta0, hu0, hv0, \
                  nx, ny, \
                  dx, dy, dt, \
                  g, f, r, A, \
                  wind_stress=Common.WindStressParams(), \
                  block_width=16, block_height=16):
-        self.cl_ctx = cl_ctx
+        #self.cl_ctx = cl_ctx # How could this work...? (Replaced by the ten lines below.)
```
Comment (babrodtk) [inline doc/reference_python/SWESimulators/CTCS.py]: This class should not create its own OpenCL context. It should receive a context to run.

File: doc/reference_python/SWESimulators/Common.py [hunk 2]
```diff
@@ -13,7 +13,7 @@ def get_kernel(cl_ctx, kernel_filename, block_width, block_height):
     #Read the proper program
     module_path = os.path.dirname(os.path.realpath(__file__))
     fullpath = os.path.join(module_path, kernel_filename)
-    options = ['-I', "'" + module_path + "'"]
+    options = ['-I', module_path]
```
Comment (babrodtk) [inline doc/reference_python/SWESimulators/Common.py]: This will not work on windows

File: mgr/src/initconditions.cpp [hunk 3]
```diff
@@ -42,8 +42,8 @@ InitConditions::InitConditions()
 
 inline Field2D generateBathymetry(int no, int nx, int ny, float width, float height)
 {
-    const float dx = width / (nx - 1);
-    const float dy = height / (ny - 1);
+    const float dx = width / nx;
```
Comment (babrodtk) [inline mgr/src/initconditions.cpp]: Why has this been changed?

File: sim/src/kernels/common.opencl [hunk 4]
```diff
@@ -0,0 +1,304 @@
+#ifndef COMMON_OPENCL
```
Comment (babrodtk) [inline sim/src/kernels/common.opencl]: It will become a problem to keep the python prototype/c++ production code coherent with this kind of copying of code. I suggest that we use the Python code as the basis, and perhaps move some of the python notebooks to make it more natural.

File: sim/src/kernels/computeEta.cl [hunk 5]
```diff
@@ -4,64 +4,87 @@
 #ifndef __OPENCL_VERSION__
 #define __kernel
```
Comment (babrodtk) [inline sim/src/kernels/computeEta.cl]: Why are these here? Should this file be compiled at all if we are not using OpenCL?

File: sim/src/kernels/computeEta.cl [hunk 6]
```diff
@@ -4,64 +4,87 @@
...
+        //Data
+        __global float* U_ptr, int U_pitch,
+        __global float* V_ptr, int V_pitch,
+        __global float* eta_ptr, int eta_pitch) {
 
-    // work item within group
-	const unsigned int lx = get_local_id(0);
-	const unsigned int ly = get_local_id(1);
+    //Index of thread within block
+    const int tx = get_local_id(0);
```
Comment (babrodtk) [inline sim/src/kernels/computeEta.cl]: Again, this kind of change makes it very difficult to keep one coherent code. Bugfixes in the python version will not propagate easily here, and vise versa.

File: sim/src/simulator.cpp [hunk 7]
```diff
@@ -331,6 +384,11 @@ void Simulator::_execNextStep(ProfileInfo *profInfo)
     pimpl->currTime += pimpl->dt; // advance simulation time
 }
 
+Field2D Simulator::_H() const
```
Comment (babrodtk) [inline sim/src/simulator.cpp]: It is (in my very strong opinion) a terrible practice to use names starting with an underscore. Please refrain from using that (it's more or less reserved by compilers for intrinsics etc. when using double underscores).

**Comment by setmar** [comment]
@babrodtk : The commit with changes to the jupyter notebooks has been dropped.
